plugins/ModelChecker/ModelChecker.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from PyQt5.QtCore import QTimer
+from cura.Scene.CuraSceneNode import CuraSceneNode
+
+from UM.Application import Application
+from UM.Tool import Tool
+from UM.Message import Message
+from UM.i18n import i18nCatalog
+from UM.Scene.Iterator.DepthFirstIterator import DepthFirstIterator
+
+catalog = i18nCatalog("cura")
+
+
+SHRINKAGE_THRESHOLD = 0.5
+WARNING_SIZE_XY = 150
+WARNING_SIZE_Z = 100
+
+
+class ModelChecker(Tool):
+    def __init__(self):
+        super().__init__()
+
+        self._last_known_tool_id = None
+
+        self._controller.activeToolChanged.connect(self._onActiveToolChanged)
+
+    def checkObjects(self, nodes_to_check):
+        warning_nodes = []
+        global_container_stack = Application.getInstance().getGlobalContainerStack()
+        if global_container_stack is None:
+            return []
+        material_shrinkage = {}
+        need_check = False
+
+        # Get all shrinkage values of materials used
+        for extruder_position, extruder in global_container_stack.extruders.items():
+            shrinkage = extruder.material.getProperty("material_shrinkage_percentage", "value")
+            if shrinkage is None:
+                shrinkage = 0
+            if shrinkage > SHRINKAGE_THRESHOLD:
+                need_check = True
+            material_shrinkage[extruder_position] = shrinkage
+
+        # Check if we can bail out fast
+        if not need_check:
+            return
+
+        # Check node material shrinkage and bounding box size
+        for node in nodes_to_check:
+            node_extruder_position = node.callDecoration("getActiveExtruderPosition")
+            if material_shrinkage[node_extruder_position] > SHRINKAGE_THRESHOLD:
+                bbox = node.getBoundingBox()
+                if bbox.width >= WARNING_SIZE_XY or bbox.depth >= WARNING_SIZE_XY or bbox.height >= WARNING_SIZE_Z:
+                    warning_nodes.append(node)
+
+        return warning_nodes
+
+    def checkAllSliceableNodes(self):
+        # Add all sliceable scene nodes to check
+        scene = Application.getInstance().getController().getScene()
+        nodes_to_check = []
+        for node in DepthFirstIterator(scene.getRoot()):
+            if node.callDecoration("isSliceable"):
+                nodes_to_check.append(node)
+        return self.checkObjects(nodes_to_check)
+
+    ##  Display warning message
+    def showWarningMessage(self, warning_nodes):
+        caution_message = Message(catalog.i18nc(
+            "@info:status",
+            "Some models may not be printed optimal due to object size and material chosen [%s].\n"
```

Comment:
Perhaps use `{models_list}` or something similar instead of the `%s` notation. Since the actual replacement key is 4 lines later in the code it's kind of hard to figure out what should be placed in that bracket.